### PR TITLE
Fixes #21317 - graceful handling of VM association error

### PR DESCRIPTION
**Pull request by iNecas**
Instead of immediate failure, just log the issue and proceed with the
import.

### CODE REVIEW

File: app/models/compute_resource.rb [hunk 1]
```diff
@@ -360,6 +360,31 @@ def image_exists?(image)
...
   end
 
+  def associate_hosts_support?
+    respond_to? :associated_host
+  end
+
+  def associate_hosts
+    ret = { :hosts => [], :fail_count => 0 }
+    return ret
+    vms(:eager_loading => true).each do |vm|
```
[the remark this answers is not in the document]
Comment (iNecas) [reply app/models/compute_resource.rb]: Ups: residual from testing. Fixed. Good dog!

**Comment by iNecas** [comment]
I've took @dLobatog's patch from https://bugzilla.redhat.com/show_bug.cgi?id=1496680#c7 and polished it a bit.

**Comment by timogoebel** [comment]
While we're touching this, how about extracting this to a Service Class? I think the result of `associate_hosts` method is pretty unintuitive and could be done better. Just a suggestion, I'm fine with this as well.

**Comment by iNecas** [comment]
Makes sense: before I jump on it, do you have some idea on:

1. name for the class
2. how it would be used from the controller?

I can think of something, but if you already have some thoughts, I can incorporate it in the patch

File: app/controllers/api/v2/compute_resources_controller.rb [hunk 2]
```diff
@@ -149,18 +149,12 @@ def available_security_groups
...
-          @compute_resource.vms(:eager_loading => true).each do |vm|
-            if Host.for_vm(@compute_resource, vm).empty?
-              host = @compute_resource.associated_host(vm)
-              if host.present?
-                host.associate!(@compute_resource, vm)
-                @hosts << host
-              end
-            end
-          end
+        if @compute_resource.associate_hosts_support?
```
Comment (timogoebel) [inline app/controllers/api/v2/compute_resources_controller.rb]: Nitpick: Perhaps `@compute_resource.supports_associate_hosts?`?

File: app/controllers/api/v2/compute_resources_controller.rb [hunk 3]
```diff
@@ -149,18 +149,12 @@ def available_security_groups
...
+          @hosts = @compute_resource.associate_hosts[:hosts]
+          render 'api/v2/hosts/index', :layout => 'api/v2/layouts/index_layout'
+        else
+          render_message(_('Associating vms is not supported for this compute resource'), :status => :unprocessable_entity)
```
Comment (timogoebel) [inline app/controllers/api/v2/compute_resources_controller.rb]: Capitalize VMs: `Associating VMs is not supported for this compute resource`?

**Review by timogoebel — COMMENTED** [review]
@iNecas. Cool. How about something like this:

```ruby
associater = ComputeResourceHostAssociater.new(
  :compute_resource => self || @compute_resource, # if you want to call it in the controller or not
).associate(vms)
if associater.errors?
  puts _("Errors during associate: %s") % associater.errors.full_messages.to_sentence
else
  puts _("%s Hosts were associated. You are awesome.") % associater.count
end
# Or even add the possibility to grab a list of successful / failed hosts so we could store them in a table in the future.
```

Definitely not the scope of this PR, but: All in all, I think this would be a good job for foreman-tasks. Can we do something similar with ActiveJob? So we can actually show a progress bar or similar visualization? At least for VMWare, this call is pretty slow.

**Comment by ares** [comment]
Switching to WoC

**Comment by iNecas** [comment]
I believe I've addressed all the comments

**Comment by timogoebel** [comment]
@iNecas: Looks good to me, thanks. Well done. Now that we have the class isolated, would you mind adding some basic test coverage? :-) Thanks.

**Comment by iNecas** [comment]
It's already implicitly covered in the v2 compute resources controller, but I can be more explicit, yes

File: app/services/compute_resource_host_associator.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+  private
+
+  def associate_vm(vm)
+    host = compute_resource.associated_host(vm)
+    if host.present?
+      host.associate!(compute_resource, vm)
+      @hosts << host
+    end
+  rescue => e
```
[the remark this answers is not in the document]
Comment (iNecas) [reply app/services/compute_resource_host_associator.rb]: I'm sorry doggy, but we don't have this cop enabled and I feel quite strongly this is the right approach for this code. Thanks for your input. Take care ☮️

**Comment by iNecas** [comment]
Tests added. Squashed. Rebased. Should be ready for final review/merge.

File: app/controllers/compute_resources_controller.rb [hunk 5]
```diff
@@ -42,19 +42,20 @@ def edit
...
-          if host.present?
-            host.associate!(@compute_resource, vm)
-            count += 1
-          end
-        end
+    associator = ComputeResourceHostAssociator.new(@compute_resource)
+    if associator.supported_compute_resource?
+      associator.associate_hosts
+      messages = []
+      messages << _('%s VMs associated to hosts') % associator.hosts.count
```
Comment (timogoebel) [inline app/controllers/compute_resources_controller.rb]: '12 VMs were each associated to a host.' ?

File: app/controllers/compute_resources_controller.rb [hunk 6]
```diff
@@ -42,19 +42,20 @@ def edit
...
+      if associator.fail_count > 0
+        messages << _('%s VMs failed while processing: check logs for more details') % associator.fail_count if associator.fail_count > 0
```
Comment (timogoebel) [inline app/controllers/compute_resources_controller.rb]: The `if associator.fail_count > 0` is redundant here.

File: app/services/compute_resource_host_associator.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (timogoebel) [reply app/services/compute_resource_host_associator.rb]: rescue "StandardError" should get rid of this.

File: app/services/compute_resource_host_associator.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    @fail_count += 1
+    logger.error "Could not associate VM #{vm}"
```
Comment (timogoebel) [inline app/services/compute_resource_host_associator.rb]: `Foreman::Logging.exception("Could not associate VM #{vm}", e)` ?

File: app/controllers/api/v2/compute_resources_controller.rb [hunk 8]
```diff
@@ -148,19 +148,15 @@ def available_security_groups
...
-          @compute_resource.vms(:eager_loading => true).each do |vm|
-            if Host.for_vm(@compute_resource, vm).empty?
-              host = @compute_resource.associated_host(vm)
-              if host.present?
-                host.associate!(@compute_resource, vm)
-                @hosts << host
-              end
-            end
-          end
+        if associator.supported_compute_resource?
```
Comment (timogoebel) [inline app/controllers/api/v2/compute_resources_controller.rb]: Nit: I think, I'd make this method part of the compute resource: `@compute_resource.supports_host_association?` or something.

File: app/services/compute_resource_host_associator.rb [hunk 9]
```diff
@@ -0,0 +1,36 @@
+class ComputeResourceHostAssociator
+  attr_accessor :compute_resource, :hosts, :fail_count
+  delegate :logger, :to => :Rails
+
+  def initialize(compute_resource)
```
Comment (timogoebel) [inline app/services/compute_resource_host_associator.rb]: This class will generate a lot of sql queries. We should probably open a ticket to make this more efficient once this is merged.

**Review by timogoebel — CHANGES_REQUESTED** [review]
@iNecas: Thanks. This looks a lot better than before! 👍
I left some comments inline.

I do think, the association isn't very efficiently implemented but this should do for now. Let's just remember to open a ticket to improve this if possible.

File: app/services/compute_resource_host_associator.rb (hunk 9 above)
Comment (iNecas) [reply app/services/compute_resource_host_associator.rb]: http://projects.theforeman.org/issues/22172

**Comment by iNecas** [comment]
All valid inputs, comments should be addressed now.

**Comment by iNecas** [comment]
All green, should be ready for squash&merge

**Review by timogoebel — APPROVED** [review]

File: app/controllers/compute_resources_controller.rb [hunk 10]
```diff
@@ -42,19 +42,20 @@ def edit
...
+      messages = []
+      messages << _('%s VMs were each associated to hosts') % associator.hosts.count
+      if associator.fail_count > 0
+        messages << _('%s VMs failed while processing: check logs for more details') % associator.fail_count
+        process_error(:error_msg => messages.join(', '))
+      else
+        process_success(:success_msg => messages.join(', '))
       end
+    else
+      process_error(:error_msg => 'Associating vms is not supported for this compute resource')
```
Comment (timogoebel) [inline app/controllers/compute_resources_controller.rb]: Nit: "Associating VMs..." (capitalize VMs)

File: app/controllers/compute_resources_controller.rb [hunk 11]
```diff
@@ -42,19 +42,20 @@ def edit
...
-          if host.present?
-            host.associate!(@compute_resource, vm)
-            count += 1
-          end
-        end
+    if @compute_resource.supports_host_association?
+      associator = ComputeResourceHostAssociator.new(@compute_resource)
+      associator.associate_hosts
+      messages = []
+      messages << _('%s VMs were each associated to hosts') % associator.hosts.count
```
Comment (timogoebel) [inline app/controllers/compute_resources_controller.rb]: This does not handle the case for 1 VM (pluralize). Also this is missing a trailing ".".
Can we add a special message when no VM at all was associated?

"No VMs matched any host in Foreman, no VMs were associated."

image

**Review by timogoebel — CHANGES_REQUESTED** [review]
@iNecas: Thanks, almost there. I found some very small things while testing this manually. If they are fixed, we're ready to 🚢.

**Comment by iNecas** [comment]
Updated, rebased, squashed

File: app/controllers/compute_resources_controller.rb [hunk 12]
```diff
@@ -42,19 +42,24 @@ def edit
...
+      associator = ComputeResourceHostAssociator.new(@compute_resource)
+      associator.associate_hosts
+      messages = []
+      if associator.hosts.empty?
+        messages << _('No VMs matched any host.')
+      else
+        messages << n_('%s VM was associated to a host.', '%s VMs were each associated to hosts.', associator.hosts.count) % associator.hosts.count
+      end
+      if associator.fail_count > 0
+        messages << _('%s VMs failed while processing: check logs for more details.') % associator.fail_count
```
Comment (timogoebel) [inline app/controllers/compute_resources_controller.rb]: This should also handle singular and plural correctly.

File: app/controllers/compute_resources_controller.rb [hunk 13]
```diff
@@ -42,19 +42,24 @@ def edit
...
+        process_error(:error_msg => messages.join(', '))
```
Comment (timogoebel) [inline app/controllers/compute_resources_controller.rb]: Now, that the messages have a dot, they cannot be joined by ', '. This will lead to '., '.

**Review by timogoebel — CHANGES_REQUESTED** [review]
@iNecas: Sorry for being a smartass here, but could you take a quick look at the two wording comments. Thanks.

File: app/controllers/compute_resources_controller.rb (hunk 13 above)
Comment (iNecas) [reply app/controllers/compute_resources_controller.rb]: Updated

File: app/controllers/compute_resources_controller.rb (hunk 12 above)
Comment (iNecas) [reply app/controllers/compute_resources_controller.rb]: Updated

**Comment by ares** [comment]
katello test failures seems unrelated

**Review by timogoebel — APPROVED** [review]

**Comment by timogoebel** [comment]
Merged, thanks @iNecas!

**Pull request merged**